### PR TITLE
Added User Info endpoint

### PR DESCRIPTION
**Pull request by fjbsantiago**
Trying to respect @ericchiang s [proposal](https://github.com/coreos/dex/issues/376#issuecomment-271725208) as best I could, I re-implemented the user info endpoint. It should work with all the connectors and was successfully tested with LDAP and OIDC connectors.

1. The AccessToken is filled with Claims, Connector Data and ConnectorID from the AuthCode.
2. It is then Serialized and sent to the user
3. The user may, then, do a request to dex's User Info endpoint using this token.

By default, when dex gets a /userinfo request, the token is deserialized, the Claims originally copied from AuthCode are extracted from it and sent back to the user as the User Info response payload.

This should work for all the connectors and was tested with the LDAP and OIDC connectors.

If the connector used to authenticate the user implements the GetUserInfo interface, then the responsibility of generating the User Info payload is delegated to it.

This is the case of the Oidc connector.
It uses the real AccessToken, previously stored as a field in the ConnectorData, to forward a genuine /userinfo request to the original IdP.
The Claims extracted from the AuthCode are used to generate the initial user info payload object, and this object is extended with whatever claims are returned by the original IdP.

Please, let me know if there is anything that can be improved.

### CODE REVIEW

**Comment by redbaron** [comment]
OMG, can't wait for it to be merged :)

**Comment by fjbsantiago** [comment]
Good to know there are more people interested in this, @redbaron !

Perhaps you could try out the latest Dex code from that repo I requested the merge from and see if you have any issues with it. 

This would probably save @ericchiang and @rithujohn191 some time and, possibly, speed up their reaction to the PR.

**Comment by ericchiang** [comment]
No user info endpoint should depend on the backing connector (https://github.com/coreos/dex/issues/376#issuecomment-339720659). We already have the username and groups to fill in the id_token, we should reuse that information to handle the userinfo endpoint.

**Comment by fjbsantiago** [comment]
Thank you for your comment @ericchiang !
I made sure that the default behavior of the userinfo endpoint is to simply return the claims like you mentioned. The same clames that were already stored in the id_token (username, email, groups).

The only deviation from this behavior, is in case you are working with the OIDC connector, in which case Dex will try to get additional claims from the actual /userinfo endpoint in the original IdP.

This makes it possible for anyone using Dex to possibly expect richer user information returned by Oidc providers connected to Dex.

File: connector/oidc/oidc.go [hunk 1]
```diff
@@ -215,10 +225,57 @@ func (c *oidcConnector) HandleCallback(s connector.Scopes, r *http.Request) (ide
 		Email:         claims.Email,
 		EmailVerified: claims.EmailVerified,
 	}
+
+	// Add AccessToken to user identity for future requests
+	connData, err := json.Marshal(connectorData{
```
Comment (ericchiang) [inline connector/oidc/oidc.go]: Connector data is private, containing data like upstream access tokens. It shouldn't be shared.

Comment (fjbsantiago) [reply connector/oidc/oidc.go]: The AccessToken can still be encrypted as per your suggestion: https://github.com/coreos/dex/issues/376#issuecomment-271811004

Comment (ericchiang) [reply connector/oidc/oidc.go]: If it's okay I'm just going to submit an implementation.

Comment (fjbsantiago) [reply connector/oidc/oidc.go]: Yes, of course.

**Comment by seanlaff** [comment]
What additional work does this PR need to pass review? Looking to help. We're trying to use dex as an generic oauth source for use with grafana, but grafana expects the `/userinfo` endpoint to exist.

**Comment by MaksymBilenko** [comment]
Hello,
Is there any chance that this PR would be merged?

**Comment by ytsando** [comment]
Hi,
How's it going?
Is there anything I can do to help me merge this PR?

**Comment by svajiraya** [comment]
ETA on the PR merge? This is quite important for Mozilla's OIDC plugin as well.

**Comment by rithujohn191** [comment]
@ericchiang I could take a second look at this PR and get it merged if you don't have the time to review it.

**Comment by ericchiang** [comment]
@rithujohn191 please feel free :)

**Comment by kalinon** [comment]
Will need to resolve conflicts, but it looks good.

**Comment by ian-axelrod** [comment]
How's this progressing? I want to add this to AWS Cognito to support LinkedIn, but I am pretty sure Cognito expects a userinfo endpoint given how it describes the auth flow.

**Pull request closed without merging**